src/main/kotlin/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/MtSharedTableBackupManager.kt

```diff
@@ -0,0 +1,456 @@
+                    gson.toJson(tenantTableMetadata.createTableRequest).byteInputStream(charset),
+                    objectMetadata)
+            s3.putObject(putObjectReq)
+        }
+    }
+
+    protected fun commitBackupMetadata(backupMetadata: MtBackupMetadata) {
+        val currentBackupMetadata = getBackup(backupMetadata.mtBackupName)
+
+        val newBackupMetadata: MtBackupMetadata
+        when (currentBackupMetadata) {
+            null -> newBackupMetadata = backupMetadata
+            else -> {
+                if (currentBackupMetadata.status != Status.IN_PROGRESS) {
+                    throw MtBackupException("${backupMetadata.mtBackupName} cannot continue, " +
+                            " status is: ${currentBackupMetadata.status}")
+                }
+                newBackupMetadata = currentBackupMetadata.merge(backupMetadata)
+            }
+        }
+
+        // write out metadata of what was backed up
+        val backupJson = gson.toJson(newBackupMetadata)
+        val backupMetadataBytes = backupJson.toByteArray(charset)
+        val objectMetadata = ObjectMetadata()
+        objectMetadata.contentLength = backupMetadataBytes.size.toLong()
+        objectMetadata.contentType = "application/json"
+        val putObjectReq = PutObjectRequest(s3BucketName, getBackupMetadataFile(newBackupMetadata.mtBackupName),
+                backupJson.byteInputStream(charset), objectMetadata)
+        s3.putObject(putObjectReq)
+    }
+
+    protected open fun createBackupData(
+        createBackupRequest: CreateBackupRequest,
+        physicalTableName: String,
+        mtDynamo: MtAmazonDynamoDbBase
+    ): MtBackupMetadata {
+        var lastRow: TenantTableRow? = null
+        val tenantTables: HashMap<TenantTableBackupMetadata, Long> = Maps.newHashMap()
+        var numPasses = 0
+        do {
+            val scanRequest: ScanRequest = ScanRequest(physicalTableName)
+                    .withExclusiveStartKey(lastRow?.attributeMap)
+            val scanResult = mtDynamo.scan(scanRequest)
+            if (scanResult.lastEvaluatedKey != null) {
+                lastRow = TenantTableRow(scanResult.lastEvaluatedKey)
+            }
+
+            val rowsPerTenant = MultimapBuilder.ListMultimapBuilder
+                    .linkedHashKeys().hashSetValues().build<TenantTable, TenantTableRow>()
+            for (i in 0..(scanResult.items.size - 1)) {
+                val row = scanResult.items[i]
+                val tenant = row.get(sharedTableMtDynamo.scanTenantKey)!!.s
+                val virtualTable = row.get(sharedTableMtDynamo.scanVirtualTableKey)!!.s
+                rowsPerTenant.put(TenantTable(tenantName = tenant, virtualTableName = virtualTable),
+                        TenantTableRow(scanResult.items[i]))
+            }
+
+            flushScanResult(createBackupRequest.backupName, numPasses, rowsPerTenant)
```

Comment:
Do we run the risk of creating many small s3 files, given the current shared table layout (tenant id & table name in hash key)?

Reply:
yes, that's a risk. I figure we can optimize that if/when it becomes a problem.